driven/generic/adapter.py

```diff
@@ -175,18 +181,250 @@ def annotate_new_metabolite(metabolite):
+                self.add_gene(feature)
+
+    def add_plasmid(self, plasmid):
+        """Add plasmid features to the initial model.
+        No plasmid instance in cameo, so changes are made in model genes and reactions directly
+
+        :param plasmid: gnomic.Plasmid
+        :return:
+        """
+        for feature in plasmid.features():
+            self.add_gene(feature)
+
+    def knockout_gene(self, feature):
+        """Perform gene knockout.
+        Use feature name as gene name
+
+        :param feature: gnomic.Feature
+        :return:
+        """
+        gene = self.model.genes.query(feature.name, attribute="name")
+        if gene:
+            gene[0].knock_out()
+            self.changes['removed']['genes'].add(gene[0])
+            logger.info('Gene knockout: {}'.format(gene[0].name))
+        else:
+            logger.info('Gene for knockout is not found: {}'.format(feature.name))
+
+    def add_gene(self, feature):
+        """Perform gene insertion.
+        Find all the reactions associated with this gene using KEGGClient and add them to the model
+
+        :param feature: gnomic.Feature
+        :return:
+        """
+        logger.info('Add gene: {}'.format(feature.name))
+        identifier = feature_id(feature)
+        if self.model.genes.query(identifier, attribute='name'):  # do not add if gene is already there
+            logger.info('Gene {} exists in the model'.format(feature.name))
+            return
+        for reaction_id, equation in self.genes_to_reactions.get(identifier, {}).items():
+            self.add_reaction(reaction_id, equation, identifier)
+        logger.info('Gene added: {}'.format(identifier))
+
+    def add_reaction(self, reaction_id, equation, gene_name):
+        """Add new reaction by rn ID from equation, where metabolites defined by kegg ids.
+
+        :param reaction_id: reaction rn ID
+        :param equation: equation string, where metabolites are defined by kegg ids
+        :param gene_name: gene name
+        :return:
+        """
+        reaction = Reaction(reaction_id)
+        self.model.add_reactions([reaction])
+        equation = map_equation_to_bigg(equation, self.compartment)
+        logger.info('New reaction: {}'.format(equation))
+        reaction.build_reaction_from_string(equation)
+        for metabolite in reaction.metabolites:
+            if metabolite.formula is None:  # unknown metabolite
+                self.annotate_new_metabolite(metabolite)
+                self.create_exchange(metabolite)
+                self.changes['added']['metabolites'].add(metabolite)
+        reaction.gene_reaction_rule = gene_name
+        self.changes['added']['reactions'].add(reaction)
+
+
+class MediumChangeModel(ModelModificationMixin):
+    """
+    Applies medium on cameo model
+    """
+
+    def __init__(self, model, medium):
+        """
+        Parameters
+        ----------
+        model
+            cameo model
+        medium
```

Comment:
The medium change model can set the correct uptake rates if you know them. It is probably more relevant than the concentration. You can add an extra key to the dictionaries in the list.